examples/pipeline/wav2letter.py

```diff
@@ -0,0 +1,1179 @@
+            x = ''.join(i for i, _ in itertools.groupby(x))
+            x = x.replace(char_blank, "")
+            # x = x.strip()
+            return x
+
+
+coder = Coder(labels)
+encode = coder.encode
+decode = coder.decode
+vocab_size = coder.length
+print("vocab_size", vocab_size, flush=True)
+
+
+# Model
+
+model = Wav2Letter(num_features, vocab_size)
+
+def model_length_function(tensor):
+    return int(tensor.shape[0])//2 + 1
+
+
+# Dataset
+
+
+class IterableMemoryCache:
+
+    def __init__(self, iterable):
+        self.iterable = iterable
+        self._iter = iter(iterable)
+        self._done = False
+        self._values = []
+
+    def __iter__(self):
+        if self._done:
+            return iter(self._values)
+        return itertools.chain(self._values, self._gen_iter())
+
+    def _gen_iter(self):
+        for new_value in self._iter:
+            self._values.append(new_value)
+            yield new_value
+        self._done = True
+
+    def __len__(self):
+        return len(self._iterable)
+
+
+class MapMemoryCache(torch.utils.data.Dataset):
+    """
+    Wrap a dataset so that, whenever a new item is returned, it is saved to memory.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+        self._cache = [None] * len(dataset)
+
+    def __getitem__(self, n):
+        if self._cache[n]:
+            return self._cache[n]
+
+        item = self.dataset[n]
+        self._cache[n] = item
+
+        return item
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+class Processed(torch.utils.data.Dataset):
+
+    def __init__(self, process_datapoint, dataset):
+        self.process_datapoint = process_datapoint
+        self.dataset = dataset
+
+    def __getitem__(self, n):
+        try:
+            item = self.dataset[n]
+            return self.process_datapoint(item)
+        except (FileNotFoundError, RuntimeError):
+            return None
+
+    def __next__(self):
+        try:
+            item = next(self.dataset)
+            return self.process_datapoint(item)
+        except (FileNotFoundError, RuntimeError):
+            return self.__next__()
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+def process_datapoint(item):
+    transformed = item[0]  # .to(device, non_blocking=non_blocking)
+    target = item[2].lower()
+
+    transformed = transforms(transformed)
+
+    transformed = transformed[0, ...].transpose(0, -1)
+
+    target = " " + target + " "
+    target = encode(target)
+    target = torch.tensor(target, dtype=torch.long, device=transformed.device)
+
+    transformed = transformed  # .to("cpu")
+    target = target  # .to("cpu")
+    return transformed, target
+
+
+def datasets_librispeech():
+
+    def create(tag):
+
+        if isinstance(tag, str):
+            data = LIBRISPEECH(
+                root, tag, folder_in_archive=folder_in_archive, download=False)
+        else:
+            data = torch.utils.data.ConcatDataset([LIBRISPEECH(
```

Comment:
This could also be done using ```sum``` since ConcatDataset can be created [via add](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataset.py#L27).

Reply:
Good point